### PR TITLE
Refactor padding calculation to improve readability.

### PR DESCRIPTION
**Pull request by eustas**
Drive-by: generalise Rect to use with ssize_t (negative origin).

### CODE REVIEW

File: lib/jxl/image.h [hunk 1]
```diff
@@ -194,21 +194,65 @@ class Image3;
...
+
+  JXL_MUST_USE_RESULT RectT Translate(int64_t x_offset,
+                                      int64_t y_offset) const {
+    return RectT(x0_ + x_offset, y0_ + y_offset, xsize_, ysize_);
+  }
+
+  T x0() const { return x0_; }
+  T y0() const { return y0_; }
+  size_t xsize() const { return xsize_; }
+  size_t ysize() const { return ysize_; }
```
Comment (veluca93) [inline lib/jxl/image.h:230]: Can we do x1, y1 while we're at it?

File: lib/jxl/image.h [hunk 2]
```diff
@@ -194,21 +194,65 @@ class Image3;
...
+  T x0_;
+  T y0_;
+
+  size_t xsize_;
+  size_t ysize_;
+};
+
+class Rect : public RectT<size_t> {
+ public:
+  constexpr Rect(size_t xbegin, size_t ybegin, size_t xsize_max,
```
Comment (veluca93) [inline lib/jxl/image.h:249]: Could do using RectT::RectT, no?

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/image.h (hunk 1 above)
Comment (eustas) [reply lib/jxl/image.h]: Sure.

File: lib/jxl/image.h (hunk 2 above)
Comment (eustas) [reply lib/jxl/image.h]: Merged back Rect and RectT. Inheritance of templatized producer without CRTP is a hell, and with it - it looks disgusting.

Comment (veluca93) [reply lib/jxl/image.h]: Just thinking out loud, but couldn't we make all rects use ssize_t?

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/image.h (hunk 2 above)
Comment (eustas) [reply lib/jxl/image.h]: We could try. One thing I foresee is problem with large images on 32-bit system... But it is likely, we will get OOM first.

**Pull request merged**